### ISSUE TITLE
Filter sites in SiteSwitchForm

### ISSUE THREAD

**Issue by rinti** [wagtail/wagtail#4637]
### Issue Summary

We have several sites with individual SiteSettings for each site, but there is no way to handle permissions as to who sees which site in the dropdown for site settings (the one on the top right corner). So basically someone on Site 1 can go edit Site 3 settings through the dropdown. (Unless I'm missing some obvious permission or wagtail setting)

The code that handles this is in [contrib/settings/forms.py](https://github.com/wagtail/wagtail/blob/master/wagtail/contrib/settings/forms.py#L18). My suggestion is adding a hook to filter sites in the `__init__` and also preferrably passing along the request object from [contrib/settings/views.py](https://github.com/wagtail/wagtail/blob/master/wagtail/contrib/settings/views.py#L88)

I'd be happy to contribute if this sounds good.

**Comment by gasman** [comment]
Hi @rinti,
The problem with making permission logic configurable through hooks is that it makes the code hard for us to maintain in future - it's hard to refactor the code (e.g. optimising queries) because we have no way of knowing what kind of customisations people have made, and what we're going to break as a result.

Giving users permission over a single site setting record is probably quite a common requirement, so it would be good to have that functionality built in to Wagtail, rather than needing a customisation. Is there some way we can extend the permission model to support that?

**Comment by rinti** [comment]
Yeah, perhaps it would be possible to change SiteSettings permissions on a per site basis. Like the document and image permissions but for each type of site setting.

So I can make a group and then under the heading "<Name of site setting> permissions" I can add which sites has read/change on that type of setting. The group admin has potential to be cluttered depending on how many types of site settings there is though.

**Comment by jasonHav** [comment]
Has anyone found a workaround for this beyond hacking the templates?

**Comment by robmoorman** [comment]
From a multisite perspective this would definitely a good feature. For now we accomplished this by monkey patching the SiteSwitchForm in combination with using threads to get the user (or request). Should be a better way.

**Comment by th3hamm0r** [comment]
With the migration to class based views in https://github.com/wagtail/wagtail/pull/11014 (Wagtail 6.0), it is now getting easier to modify the behavior.
I think the missing piece to allow an easy customization of the listed sites based on the current request without duplicating much of the code, is to extract the [querying of the sites](https://github.com/wagtail/wagtail/blob/538365fcfc5ea9135663a1b8c2511c499910e419/wagtail/contrib/settings/views.py#L119-L121) into a separate view method or (cached-)property, and then pass the resulting list to the `SiteSwitchForm` (avoiding the extra query [here](https://github.com/wagtail/wagtail/blob/538365fcfc5ea9135663a1b8c2511c499910e419/wagtail/contrib/settings/forms.py#L31)).
This alone would make it easy for devs to restrict the access based on other factors.

Independently of the above (API-)improvement, we could change the default behavior and only show those settings to the user, where the user has access to pages below the root page of the settings' site. We could use the `path` of the site's root page to check if a user has access (checking if the root page's path is a prefix of the accessible pages).
To still allow the current behavior, we may then add an additional permission to `BaseSiteSetting` (and therefore to every subclass), which allows to give access to all settings, independently of the accessible pages.
I know, this approach is not as flexible as the approach of "Giving users permission over a single site setting record", but it should be quite simple to implement, and I currently cannot think of cases of past projects, where that is not enough.
